
## A project brief is :
This dataset collects information from 100k medical appointments in Brazil and is focused on the question of whether or not patients show up for their appointment. 110.527 medical appointments with 14 features

## Import Necessary libraries




In [ ]:
import pandas as pd
import numpy as np


## Questions:

1- Does Show rate differs by gender? males may have higher probability of show up due to social norms that constricts the movement of women?

2- How the scholarship affects the probability of show up? those who were eligible for the scholarship may have shown up more often because the are able to afford the costs or because the fear of loosing the scholarship?

3- AppointmentDay: How delay affects show up?

## We can find answers of all questions in the data let us find the answes!

In [ ]:
df = pd.read_csv("KaggleV2-May-2016.csv") 

In [ ]:
# to explore data we have 
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [ ]:
# to desplay name of columns only to remove aunecessary column
print(list(df.columns.values))
print(len(list(df.columns.values)))

['PatientId', 'AppointmentID', 'Gender', 'ScheduledDay', 'AppointmentDay', 'Age', 'Neighbourhood', 'Scholarship', 'Hipertension', 'Diabetes', 'Alcoholism', 'Handcap', 'SMS_received', 'No-show']
14


In [ ]:
#display number of rows and columns
df.shape

(110527, 14)

In [ ]:
# To display : column labels, data types, and the number of cells in each column.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


# There is no missing values in our data.

### As we can see the data types needs to be fixed at some columns:
- PatientId and AppointmentID are currently numeric .
where in fact they have should been string or object as they have no numeric meaning.
- ScheduledDay and AppointmentDay should be converted from object to date-time
- No-show should be recoded and converted into integer.
-ScheduledDay and AppointmentDay: for more simplicity, we might keep only the date and delete the time.

### And create some columns:
- Age column to a categorical variable in order to clearly see of ages "Age_categoties".
- A new variable for integers waiting days "Wait_days".
- A new variant for Diabetics and Hypertension also to detect patients suffering from both diseases "diabetic_hiper".



In [ ]:
# 'PatientId' and 'AppointmentID' from float to string
df["PatientId"] = df['PatientId'].astype(str)
df["AppointmentID"] = df['AppointmentID'].astype(str)

In [ ]:
# 'ScheduledDay' and 'AppointmentDay' to datetime type.
df['ScheduledDay'] =  pd.to_datetime(df['ScheduledDay']).dt.date
df['AppointmentDay'] =  pd.to_datetime(df['AppointmentDay']).dt.date

In [ ]:
# To rename column 'No-show' to 'show' to prevent any misconciption
df.rename(columns = {'No-show': 'show'}, inplace = True) 

In [ ]:
# To convert 'show' to False = '1' & True = '0'.
df['show'].replace('Yes', 0,inplace=True)
df['show'].replace('No', 1,inplace=True)

In [ ]:
df['show'].unique()

array([1, 0])

In [ ]:
# To rename column 'gender' to 'male' to prevent any misconciption
df.rename(columns = {'Gender': 'male'}, inplace = True) 

In [ ]:
# To convert 'male' to male = '1' & female = '0'.
df['male'].replace('F', 0,inplace=True)
df['male'].replace('M', 1,inplace=True)

# To check our solution


In [ ]:
df['male'].unique()

array([0, 1])

In [ ]:
df['show'].unique()

array([1, 0])

In [ ]:
# To check what happen.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   PatientId       110527 non-null  object
 1   AppointmentID   110527 non-null  object
 2   male            110527 non-null  int64 
 3   ScheduledDay    110527 non-null  object
 4   AppointmentDay  110527 non-null  object
 5   Age             110527 non-null  int64 
 6   Neighbourhood   110527 non-null  object
 7   Scholarship     110527 non-null  int64 
 8   Hipertension    110527 non-null  int64 
 9   Diabetes        110527 non-null  int64 
 10  Alcoholism      110527 non-null  int64 
 11  Handcap         110527 non-null  int64 
 12  SMS_received    110527 non-null  int64 
 13  show            110527 non-null  int64 
dtypes: int64(9), object(5)
memory usage: 11.8+ MB


In [ ]:
df.head()

PatientId AppointmentID  male ScheduledDay AppointmentDay  Age  \
0   29872499824296.0       5642903     0   2016-04-29     2016-04-29   62   
1  558997776694438.0       5642503     1   2016-04-29     2016-04-29   56   
2    4262962299951.0       5642549     0   2016-04-29     2016-04-29   62   
3     867951213174.0       5642828     0   2016-04-29     2016-04-29    8   
4    8841186448183.0       5642494     0   2016-04-29     2016-04-29   56   

       Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  \
0    JARDIM DA PENHA            0             1         0           0   
1    JARDIM DA PENHA            0             0         0           0   
2      MATA DA PRAIA            0             0         0           0   
3  PONTAL DE CAMBURI            0             0         0           0   
4    JARDIM DA PENHA            0             1         1           0   

   Handcap  SMS_received  show  
0        0             0     1  
1        0             0     1  
2        0             0     1  
3        0             0     1  
4        0             0     1

# All of data quailty issuses are solved, lets now check for duplicates



In [ ]:
df.duplicated().sum()

0

No Duplicate rows

 #### Change age to a categorical variable in order to clearly see of ages


In [ ]:
df["Age_categoties"] = pd.qcut(df.Age, 5, labels = ["0 - 12", "13 - 29", "30 - 44", "45 - 58", "59 - 115"])

 ## Create a new variant for Diabetics and Hypertension also to detect patients suffering from both diseases

In [ ]:
df["diabetic_hiper"] = df["Diabetes"] * df["Hipertension"]

##  Create a new variable for integers waiting days


In [ ]:
df["Wait_days"] = df["AppointmentDay"] - df["ScheduledDay"]
df["Wait_days"] = df["Wait_days"].astype(str)
df["Wait_days"] = df["Wait_days"].apply(lambda x: x.split(" ")[0])
df["Wait_days"] = df["Wait_days"].astype(int)

### With reason, age cannot be negative, row whith ngative value should be removied.

In [ ]:
df = df[df.Age != -1]


In [ ]:
pd.DataFrame({'min' : df.min(),
              'max': df.max()})

min               max
PatientId       11111462625267.0  99999967293459.0
AppointmentID            5030230           5790484
male                           0                 1
ScheduledDay          2015-11-10        2016-06-08
AppointmentDay        2016-04-29        2016-06-08
Age                            0               115
Neighbourhood          AEROPORTO        VILA RUBIM
Scholarship                    0                 1
Hipertension                   0                 1
Diabetes                       0                 1
Alcoholism                     0                 1
Handcap                        0                 4
SMS_received                   0                 1
show                           0                 1
Age_categoties            0 - 12          59 - 115
diabetic_hiper                 0                 1
Wait_days                     -6               179

### With reason, wait-days cannot be negative, row whith ngative value should be removied.

In [ ]:
df = df[df.Wait_days > 0]


 #### As we did in age column, Change wait-days to a categorical variable in order to clearly see of wait-days.

In [ ]:
#Converting wait days into wait bins
df["Wait_days"] = pd.cut(df["Wait_days"], 10, labels = ["0-17", "18-34", "35-53", "54-71", "72-89", "90-106", "107-124", "125-142", "143-160", "161-179"])

<ipython-input-54-5d983176315d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Wait_days"] = pd.cut(df["Wait_days"], 10, labels = ["0-17", "18-34", "35-53", "54-71", "72-89", "90-106", "107-124", "125-142", "143-160", "161-179"])


In [ ]:
# Summary Statistics
df.describe()

male           Age   Scholarship  Hipertension      Diabetes  \
count  71959.000000  71959.000000  71959.000000  71959.000000  71959.000000   
mean       0.331981     38.502564      0.092706      0.208897      0.074723   
std        0.470927     22.925421      0.290021      0.406523      0.262946   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000     19.000000      0.000000      0.000000      0.000000   
50%        0.000000     39.000000      0.000000      0.000000      0.000000   
75%        1.000000     57.000000      0.000000      0.000000      0.000000   
max        1.000000    115.000000      1.000000      1.000000      1.000000   

         Alcoholism       Handcap  SMS_received          show  diabetic_hiper  
count  71959.000000  71959.000000  71959.000000  71959.000000    71959.000000  
mean       0.025320      0.020025      0.493086      0.714810        0.061104  
std        0.157096      0.154072      0.499956      0.451508        0.239523  
min        0.000000      0.000000      0.000000      0.000000        0.000000  
25%        0.000000      0.000000      0.000000      0.000000        0.000000  
50%        0.000000      0.000000      0.000000      1.000000        0.000000  
75%        0.000000      0.000000      1.000000      1.000000        0.000000  
max        1.000000      4.000000      1.000000      1.000000        1.000000

### Main Remarks:

The mean age of all patients is 38.5

For all the patients:

- 9.2 percent have the scholarship
- 33 percent are males
- 67 precent are femals
- 20 percent have high blood pressure
- 7.4 percent are diabetic
- 2.5 percent drink alcohol
- 2 percent are handicapped
- 49 percent received a sms
- Show up rate is 71.9 percent
- 6.1 percent are diabetic and also hipertension

In [ ]:
# To save dataframe as CSV.
df.to_csv("Medical_Appointment_No_Shows.csv")

In [ ]:
df.head()

PatientId AppointmentID  male ScheduledDay AppointmentDay  Age  \
5    95985133231274.0       5626772     0   2016-04-27     2016-04-29   76   
6   733688164476661.0       5630279     0   2016-04-27     2016-04-29   23   
7     3449833394123.0       5630575     0   2016-04-27     2016-04-29   39   
9    78124564369297.0       5629123     0   2016-04-27     2016-04-29   19   
10  734536231958495.0       5630213     0   2016-04-27     2016-04-29   30   

     Neighbourhood  Scholarship  Hipertension  Diabetes  Alcoholism  Handcap  \
5        REPÚBLICA            0             1         0           0        0   
6       GOIABEIRAS            0             0         0           0        0   
7       GOIABEIRAS            0             0         0           0        0   
9        CONQUISTA            0             0         0           0        0   
10  NOVA PALESTINA            0             0         0           0        0   

    SMS_received  show Age_categoties  diabetic_hiper Wait_days  
5              0     1       59 - 115               0      0-17  
6              0     0        13 - 29               0      0-17  
7              0     0        30 - 44               0      0-17  
9              0     1        13 - 29               0      0-17  
10             0     1        13 - 29               0      0-17